### PR TITLE
Get generic signature of fields entered after erasure from their accessor 

### LINKED ISSUE #17069: Missing generic type in JVM signature of public field
## Compiler version

3.2.2

## Minimized code

```Scala
//> using scala "3.2.2"
//> using repository "https://maven.atlassian.com/repository/public"
//> using lib "com.atlassian.bitbucket.server:bitbucket-rest-model:7.14.0"

import com.atlassian.bitbucket.rest.repository.RestRepository
import scala.beans.BeanProperty

case class SomeRequest(
    @BeanProperty var repos: java.util.List[RestRepository]
)
```

## Output
Partial output of `scalap -c -p`
```scala
// public class generic_java_type$u002Esc$SomeRequest implements scala.Product,java.io.Serializable {
//   private java.util.List repos;
//   public java.util.List<com.atlassian.bitbucket.rest.repository.RestRepository> repos();
//     Code:
//        0: aload_0
//        1: getfield      #39                 // Field repos:Ljava/util/List;
//        4: areturn```
```
## Expectation
Scala 2 and Java add the generic type of the java.util.List to the repos field. Changing the version `//> scala 2` in the code above produces:

```scala
// public class generic_java_type$u002Esc$SomeRequest implements scala.Product,java.io.Serializable {
//   private java.util.List<com.atlassian.bitbucket.rest.repository.RestRepository> repos;
//   public java.util.List<com.atlassian.bitbucket.rest.repository.RestRepository> repos();
//     Code:
//        0: aload_0
//        1: getfield      #28                 // Field repos:Ljava/util/List;
//        4: areturn
```

We got hit by this when Jackson was unable to deserialize the class correctly because it could not reflect on the type of the list.

### PR DESCRIPTION
**Pull request by EugeneFlesselle**
Fix #17069 

@mbovel

### CODE REVIEW

File: tests/generic-java-signatures/17069.scala [hunk 1]
```diff
@@ -0,0 +1,8 @@
+
+class Foo:
+  val generic: List[String] = ???
+
+@main def Test =
+  val tpe = classOf[Foo].getDeclaredField("generic").getGenericType()
+  val tpe2 = classOf[Foo].getDeclaredMethod("generic").getGenericReturnType()
+  assert(tpe == tpe2)
```
Comment (sjrd) [inline tests/generic-java-signatures/17069.scala:8]: This test will still pass if *both* the field and getter have broken signatures. We should also actually test the structure of `tpe`.

**Review by sjrd — CHANGES_REQUESTED** [review]

**Comment by sjrd** [comment]
LGTM. Would you mind squashing all the commits in a single one? The individual commits here don't bring any value for the future user of `git blame`.

**Review by sjrd — APPROVED** [review]

**Comment by som-snytt** [comment]
@sjrd thank-you for encouraging squashing, as it improved my bisecting experience.

I suggested on the linked ticket that this might be backportable but I haven't tried it. Maybe I can just try it.

**Comment by SethTisue** [comment]
Did somebody think it would be hard to backport? Or was it considered unwise to mess with generic signatures in a point release?

**Comment by EugeneFlesselle** [comment]
> Did somebody think it would be hard to backport? Or was it considered unwise to mess with generic signatures in a point release?

\cc@hamzaremmal might be informed regarding this

**Pull request merged**